netmon/netmon.go

```diff
@@ -0,0 +1,613 @@
+	n.netHandler.Delete()
+	close(n.linkDoneCh)
+	close(n.rtDoneCh)
+}
+
+func (n *netmon) logger() *logrus.Entry {
+	fields := logrus.Fields{
+		"name":   netmonName,
+		"pid":    os.Getpid(),
+		"source": "netmon",
+	}
+
+	if n.sandboxID != "" {
+		fields["sandbox"] = n.sandboxID
+	}
+
+	return netmonLog.WithFields(fields)
+}
+
+func (n *netmon) setupLogger() error {
+	level, err := logrus.ParseLevel(n.logLevel)
+	if err != nil {
+		return err
+	}
+
+	netmonLog.SetLevel(level)
+
+	netmonLog.Formatter = &logrus.TextFormatter{TimestampFormat: time.RFC3339Nano}
+
+	hook, err := lSyslog.NewSyslogHook("", "", syslog.LOG_INFO|syslog.LOG_USER, netmonName)
+	if err != nil {
+		return err
+	}
+
+	netmonLog.AddHook(hook)
+
+	announceFields := logrus.Fields{
+		"runtime-path": n.runtimePath,
+		"debug":        n.debug,
+		"log-level":    n.logLevel,
+	}
+
+	n.logger().WithFields(announceFields).Info("announce")
+
+	return nil
+}
+
+func (n *netmon) listenNetlinkEvents() error {
+	if err := netlink.LinkSubscribe(n.linkUpdateCh, n.linkDoneCh); err != nil {
+		return err
+	}
+
+	return netlink.RouteSubscribe(n.rtUpdateCh, n.rtDoneCh)
+}
+
+func convertInterface(linkAttrs *netlink.LinkAttrs, addrs []netlink.Addr) Interface {
+	if linkAttrs == nil {
+		netmonLog.Warn("Link attributes are nil")
+		return Interface{}
+	}
+
+	var ipAddrs []*IPAddress
+
+	for _, addr := range addrs {
+		if addr.IPNet == nil {
+			continue
+		}
+
+		netMask, _ := addr.Mask.Size()
+
+		ipAddr := &IPAddress{
+			Family:  IPFamily(netlinkFamily),
+			Address: addr.IP.String(),
+			Mask:    fmt.Sprintf("%d", netMask),
+		}
+
+		ipAddrs = append(ipAddrs, ipAddr)
+	}
+
+	return Interface{
+		Device:      linkAttrs.Name,
+		Name:        linkAttrs.Name,
+		IPAddresses: ipAddrs,
+		Mtu:         uint64(linkAttrs.MTU),
+		HwAddr:      linkAttrs.HardwareAddr.String(),
+	}
+}
+
+func convertRoutes(netRoutes []netlink.Route) []Route {
+	var routes []Route
+
+	for _, netRoute := range netRoutes {
+		dst := ""
+		if netRoute.Dst != nil && netRoute.Dst.IP.To4() != nil {
+			dst = netRoute.Dst.IP.String()
+		}
+
+		src := ""
+		if netRoute.Src.To4() != nil {
+			src = netRoute.Src.String()
+		}
+
+		gw := ""
+		if netRoute.Gw.To4() != nil {
+			gw = netRoute.Gw.String()
+		}
+
+		dev := ""
+		iface, err := net.InterfaceByIndex(netRoute.LinkIndex)
+		if err == nil {
+			dev = iface.Name
+		}
+
+		route := Route{
+			Dest:    dst,
+			Gateway: gw,
+			Device:  dev,
+			Source:  src,
+			Scope:   uint32(netRoute.Scope),
+		}
+
+		routes = append(routes, route)
+	}
+
+	return routes
+}
+
+func (n *netmon) scanNetwork() error {
+	links, err := n.netHandler.LinkList()
+	if err != nil {
+		return err
+	}
+
+	for _, link := range links {
+		addrs, err := n.netHandler.AddrList(link, netlinkFamily)
+		if err != nil {
+			return err
+		}
+
+		linkAttrs := link.Attrs()
+		if linkAttrs == nil {
+			continue
+		}
+
+		iface := convertInterface(linkAttrs, addrs)
+		n.netIfaces[linkAttrs.Index] = iface
```

Comment:
Worth adding a few more log calls for example when an interface is "scanned" / route converted / interface converted?